design-proposals/klip-6-ksql-testing-tool.md

```diff
@@ -0,0 +1,186 @@
+
+- _name_ (required): The name of the test case.
+description (optional): A description of what the test case is testing.
+- _format_ (optional): An array of multiple different formats to run the test case as, e.g. AVRO, JSON, DELIMITED.
+- _statements_ (required): The list of statements to execute as this test case.
+- _properties_ (optional): A map of property name to value. Can contain any valid Ksql config. The config is passed to the engine when executing the statements in the test case
+- _topics_ (optional): An array of the topics this test case needs. Allows more information about the topic to be supplied, e.g. an existing Avro schema.
+- _inputs_ (required if `expectedException` not supplied): The set of input messages to be produced to Kafka topic(s).
+- _outputs_ (required if `expectedException` not supplied) The set of output messages expected in the output topic(s).
+- _expectedException_ (required if `inputs` and `outputs` not supplied) The exception that should be thrown when executing the supplied statements.
+
+
+For more details on the test file format click ([here](https://github.com/confluentinc/ksql/blob/master/ksql-engine/src/test/resources/query-validation-tests/README.md)).
+
+
+### Testing Tool
+
+The testing tool will be a command line tool called ksql-testing-tool that accepts a test file as input parameter. The following is the way we can use the tool:
+
+```shell
+$ ksql-testing-tool /path/to/the/testfile.json
+```
+The tool will run the tests from the test file and if all the tests pass will terminate successfully printing the success message in the terminal! If any of the tests fail the tool will print out the reason for the failure in the terminal. The following is an example of successful test:
+
+```shell
+$ ksql-testing-tool /path/to/passing_test.json
+All tests passed!
+$
+```
+The following is a sample output of the tool when a test fails:
+
+```shell
+$ ksql-testing-tool /path/to/failing_test.json
+Exception in thread "main" java.lang.AssertionError: TestCase name: project-filter - project and filter in file: /path/to/correct_test.json failed while processing output row 0 topic: S1 due to: Expected <1010, 101> with timestamp=0 but was <101, 101> with timestamp=0
+  at io.confluent.ksql.test.commons.TestCase.verifyOutput(TestCase.java:190)
+  at io.confluent.ksql.testingtool.TestRunner.shouldBuildAndExecuteQuery(TestRunner.java:134)
+  at io.confluent.ksql.testingtool.TestRunner.main(TestRunner.java:96)
+Caused by: java.lang.AssertionError: Expected <1010, 101> with timestamp=0 but was <101, 101> with timestamp=0
+  at org.apache.kafka.streams.test.OutputVerifier.compareKeyValueTimestamp(OutputVerifier.java:208)
+  at io.confluent.ksql.test.commons.TestCase.verifyOutput(TestCase.java:177)
+  ... 2 more
+$
+```
+
+As you can see, the error message will include the failure reason.
+
+## Alternative Public API
```

Comment:
Personally, I thin the alternative design, with two files, needs more thought

To be useful to users I think they'd want to have directories full of json messages, where each file contains the contents of one Kafka message.

The test file should then allow the user to supply the inputs and expected outputs directories, rather than inline in the file.

The benefit of this is that tests can share the same data sets, rather than having to have them duplicated.

IMHO, we should go with the first design, i.e. the same as QTT, and then extend to allow for inputs and outputs to be defined in external locations.

Reply:
> The benefit of this is that tests can share the same data sets, rather than having to have them duplicated.

I think a major benefit this approach is not having to duplicate your statements and keep your queries and tests in sync if your primary use case is headless mode.

Reply:
> where each file contains the contents of one Kafka message
I think what's more common is a file with each line representing a record.

> allow the user to supply the inputs and expected outputs directories
Just specifying a directory wouldn't let the user control for ordering though.